lib/vagrant-libvirt/action/shutdown_domain.rb

```diff
@@ -0,0 +1,49 @@
+require 'log4r'
+
+module VagrantPlugins
+  module ProviderLibvirt
+    module Action
+      # Shutdown the domain.
+      class ShutdownDomain
+        def initialize(app, _env, target_state, source_state)
+          @logger = Log4r::Logger.new('vagrant_libvirt::action::shutdown_domain')
+          @target_state = target_state
+          @source_state = source_state
+          @app = app
+        end
+
+        def call(env)
+          timeout = env[:machine].config.vm.graceful_halt_timeout
+
+          start_time = Time.now
+
+          # call nested action first under the assumption it should try to
+          # handle shutdown via client capabilities
+          @app.call(env)
```

Comment:
@chrisroberts I've done some reworking and testing, and part of me is wondering whether instead of calling the next action in the chain, it would be better if this action either subclassed or called the GracefulHalt action directly at this point instead. As I don't think it's necessarily valid to treat them as separate actions where it's feasible to add this action elsewhere as it always needs to understand what it is wrapping. Thoughts?

Reply:
Yeah, the way it's structured now, the `ShutdownDomain` action is making an assumption/requirement on what is being done after it and if that's the case then just calling it directly would be better simply for ensuring the expected behavior. You can remove that assumed dependency, though, using the `EnvSet` action to just log the start time for you. Something like:

```ruby
b3.use Call, Message, "Attempting nice shutdowns..." do |_, b4|
  b4.use EnvSet, :libvirt_shutdown_start, Time.now
  b4.use GracefulHalt, :shutoff, :running
  b4.use ShutdownDomain, :shutoff, :running
end
```

Then, in this `ShutdownDomain` action, your `start_time` value can be:

```ruby
start_time = env.fetch(:libvirt_shutdown_start, Time.now)
```

so if the start time was set, you can use that, otherwise you just assume you have the full timeout to proceed.